[DataViews] Enhance filter component styles

chihsuan opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to WOOA7S-62

## Proposed Changes

* Update the data view filters component to match the design.
  - Use minimal styles for operator select control.
  - Truncate long filter names with ellipses
  - Use radio control for singleSelection type filter (Figma: hdK9XLiOefugy2bxalKBN1-fi-2795_100663)
  - Use checkbox control for multiSelection type filter (Figma: hdK9XLiOefugy2bxalKBN1-fi-2795_101600)
* Use the default WP body font in the storybook styles to make it consistent with the actual styles in the app.


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Make the data view filters component consistent with the design so we can use it in the Woo Analytics and other places without additional styling.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `yarn dataviews:storybook:start`
* Go to http://localhost:57863/?path=/story/dataviews-dataviews--default
* Click the filters button to open the filters panel.
* Choose different filters and see if the styles are correct. (Figma: hdK9XLiOefugy2bxalKBN1-fi-2795_90514)
* Change the filter name to a long name and very it doesn't wrap. 
* Change [this line](https://github.com/WordPress/gutenberg/blob/accde16096464ac832fbe8e249a3b61535800800/packages/dataviews/src/components/dataviews-filters/search-widget.tsx#L322) to render `ComboboxList` instead of `ListBox` to see if the combobox list is also working.


Before:

<img width="496" alt="Screenshot 2025-05-02 at 18 41 25" src="https://github.com/user-attachments/assets/325c3797-6127-4be7-9c41-1f08418afb8e" />


<img width="373" alt="Screenshot 2025-05-02 at 18 50 12" src="https://github.com/user-attachments/assets/d925e55c-a6fc-473d-86c7-44bbc1619c6b" />


After

<img width="514" alt="Screenshot 2025-05-02 at 18 37 33" src="https://github.com/user-attachments/assets/0c0095c2-9fc4-40d7-88bd-2a964187dc84" />
<img width="499" alt="Screenshot 2025-05-02 at 18 37 26" src="https://github.com/user-attachments/assets/4d9c5756-3755-45aa-baa5-a561fb38a17f" />
<img width="442" alt="Screenshot 2025-05-02 at 18 36 36" src="https://github.com/user-attachments/assets/b615aa12-4a14-4de4-af30-3510923a3a71" />



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
Add body font styling to Storybook styles

chihsuan committed:
Update OperatorSelector component to use minimal variant for improved styling

chihsuan committed:
Enhance SearchWidget component with RadioControl and CheckboxControl for improved filter selection

chihsuan committed:
Update styles in Dataviews Filters component to enhance layout and text handling

- Adjust padding in summary operators container for improved spacing
- Add text overflow handling to filter name for better UI consistency
- Modify radio and checkbox label styles for improved accessibility and interaction

chihsuan committed:
Add filterBy property to fields in fixtures for enhanced filtering options

chihsuan committed:
Remove unused variable and import

chihsuan commented:
<!-- linear:replyTo:2847520674 -->

Thanks for the review and feedback! Jay.

> I think we need some technical feedback from the components team about the use of checkbox/radio inputs within the listbox. Generally the individual option elements are expected to be single interactive units.<br>Instead of using inputs, maybe we style non-interactive elements to look like inputs?  Good point!

I agree that using non-interactive elements would be a better approach for accessibility. I'll update the implementation accordingly. 🙂 

> I think it would be beneficial to mock up the intended designs for all `option` states, for both multi-select (checkbox) and single-select (radio) listboxes. 

@<!-- linear:userMention:882ce9d9-ab24-45bb-8e9e-186735b72984 -->veronica.fasulo  If we have a mockup for all option states, could you share it here? 🙏 I'll double-check the styles and make sure they're applied correctly. Thanks!

chihsuan committed:
Replace RadioControl and CheckboxControl with custom implementations

- Introduce custom Checkbox and Radio components using clsx for dynamic class handling
- Update styles for radio and checkbox elements to enhance visual feedback and accessibility
- Remove unused imports and variables for cleaner code

chihsuan commented:
<!-- linear:isThreadRoot -->

I've replaced RadioControl and CheckboxControl with custom non-interactive implementations in in daf756917dbc8ac9f892ee01fb88213d4baa2775.

[Screen Recording 2025-05-05 at 15.39.30.mov](https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/7043620f-f543-4e4c-8404-14c208b0aca2/99997f39-c154-4ae1-8614-04558e88e415?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi83MDQzNjIwZi1mNTQzLTRlNGMtODQwNC0xNGMyMDhiMGFjYTIvOTk5OTdmMzktYzE1NC00YWUxLTg2MTQtMDQ1NThlODhlNDE1IiwiaWF0IjoxNzQ2NDMxMDQ3LCJleHAiOjMzMzE2OTkxMDQ3fQ.bFz9Mv4wM5WxpEewiJkjYd4nGqUP9xZ5e3aDELHDBaE)

chihsuan committed:
Fix check classname

youknowriad commented:
While I'm ok shipping this as interim step, I think there's a discussion to be had about whether "listbox" is the right thing rather than just considering the dropdown as a regular form with any form controls. There's also a world where the same controls from all the filters should also be rendered in a sidebar (I know there were design explorations like that). Maybe something for a separate issue.

youknowriad commented:
@oandregal Any thoughts here?

chihsuan commented:
@jameskoster 👋 


I've checked with @verofasulo (p1746587859135169/1746587662.284489-slack-C06FSDTSN82) and we've confirmed that implementing hover state changes for checkbox/radio options is not required - specifically, we don't need background color changes on hover.

So, it would be like this:


https://github.com/user-attachments/assets/587005e3-41fa-4dbd-b132-9872dae662e7


What do you think about this?

jameskoster commented:
<!-- linear:replyTo:2850169950 -->

@chihsuan that looks pretty good to me. My one suggestion would be to somehow indicate that the entire row is interactive. That could be a hover style, or just changing the cursor to `pointer`. 

How does the appearance look as you navigate via keyboard? We need to highlight the 'focussed' option some how. We could probably use the aforementioned hover style to handle that.

jameskoster commented:
> I think there's a discussion to be had about whether "listbox" is the right thing rather than just considering the dropdown as a regular form with any form controls

From the design perspective I'm leaning slightly toward using `listbox` over forms because it offers a stronger UX from an accessibility standpoint, and it's more flexible for enhancements like building a `combobox`. Would this not be compatible with the sidebar filter design?

youknowriad commented:
My thinking is more global. Filters are forms, some will have date range picker, others will have select box, some will have radio control, others multiple checkboxes. `listbox` is fine in a form, it's just a select box with images but in that case, if we want to use a list box, I don't think we should be imitating the look and feel of checkboxes and radioboxes. Just my 2cts.

chihsuan commented:
> that looks pretty good to me. My one suggestion would be to somehow indicate that the entire row is interactive. That could be a hover style, or just changing the cursor to `pointer`.

Changing the cursor to `pointer` makes sense to me. cc @verofasulo What do you think?

> How does the appearance look as you navigate via keyboard? We need to highlight the 'focussed' option some how. We could probably use the aforementioned hover style to handle that.

Yeah, we need to highlight the focused option somehow. We could maintain the current focus styles for keyboard navigation to ensure accessibility.

https://github.com/user-attachments/assets/b0fdb55b-a135-42c8-8447-e2d5df5b9c7f

jameskoster commented:
<!-- linear:replyTo:2850169950 -->

It's nit-picking territory but there's something slightly off about the selected checkbox blending into the highlight style. Could it make sense to set the selected checkbox background to `--wp-components-color-accent-darker-10` when the option is highlighted? That way the checkbox footprint remains visible.

jasmussen commented:
Good nitpick, though. 🤔 

A darkening can work, yes, especially in absence of a larger revisit of the "selected item" style, where the total inversion feels fairly hefty.

jameskoster commented:
<!-- linear:replyTo:2850169950 -->

> if we want to use a list box, I don't think we should be imitating the look and feel of checkboxes and radioboxes

It may be important to unpack that as it's basically the whole point of the PR 😀 Noting that @ciampo surfaced similar concerns in [https://github.com/WordPress/gutenberg/issues/65801#issuecomment-2388739111](https://github.com/WordPress/gutenberg/issues/65801#issuecomment-2388739111).<br><br>I don't have a strong opinion on this one. I see the UX argument (feel free to correct me on this @<!-- linear:userMention:882ce9d9-ab24-45bb-8e9e-186735b72984 -->veronica.fasulo); without a visual cue it's not clear whether a listbox is single or multi-select. Using checkbox/radio appearance provides this clarity. But I acknowledge it may also be confusing because we're presenting two (subtly) different behaviors in an identical way.

oandregal commented:
I've noticed this introduces horizontal scroll with long field labels:

https://github.com/user-attachments/assets/46a66209-d431-4404-bb1b-e9b3d7018fba

oandregal commented:
It maintains the same keyboard behavior (space key has press behavior, navigate through the items with the arrows keys, focus is contained to the list until you tab, etc.). That's good.

Just noting this will affect every product: so, style-wise, as long as is design-blessed, it's good for me. I do appreciate the ability to understand what'll happen before I interact with the widget (single vs multi selection), to be honest.

It feels weird that we are making them _look like_ the real components. I understand we don't want to use the real ones because we are using the listbox instead of a form — is that hinting at something we should address? I don't have a formed opinion on this just yet. I'd like to hear @ciampo @mirka thoughts before landing this PR.

ciampo commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -81,6 +75,30 @@ function generateFilterElementCompositeItemId(
 	return `${ prefix }-${ filterElementValue }`;
 }
 
+const Checkbox = ( { checked }: { checked: boolean } ) => {
+	return (
+		<span
+			className={ clsx(
+				'dataviews-filters__search-widget-listitem-checkbox',
+				{ 'is-checked': checked }
+			) }
+		>
+			{ checked && <Icon icon={ check } /> }
+		</span>
+	);
+};
+
+const Radio = ( { checked }: { checked: boolean } ) => {
+	return (
+		<span
+			className={ clsx(
+				'dataviews-filters__search-widget-listitem-radio',
+				{ 'is-checked': checked }
+			) }
+		/>
+	);
+};
+
```
Calling these two components "checkbox" and "radio" implies the wrong abstraction and semantics ([Checkbox](https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/) and [Radio](https://www.w3.org/WAI/ARIA/apg/patterns/radio/) semantics are different from [listbox semantics](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) ).

These should likely be called something like `SingleSelectionOption` and `MultiSelectionOption`, or something similar, and have a `selected` prop (instead of `checked`). Classnames should follow similar naming, too.

chihsuan replied:
Thanks for the feedback! I've updated the code to use the suggested naming and semantics in 76617fcac6e7b7310d9f6455690edcf5f0bdd5a3.

chihsuan committed:
Update styles in Dataviews Filters component to improve hover and checked states

- Refactor hover and focus styles to use data-focus-visible for better accessibility
- Change background and border colors for checked state to darker variants for enhanced visibility

chihsuan committed:
Refactor SearchWidget component to improve selection handling

- Rename Checkbox and Radio components to MultiSelectionOption and SingleSelectionOption for clarity
- Update class names in styles to reflect new component names and improve semantic meaning
- Adjust selection logic to use 'selected' prop for better consistency in multi and single selection scenarios

chihsuan commented:
> Could it make sense to set the selected checkbox background to `--wp-components-color-accent-darker-10` when the option is highlighted? That way the checkbox footprint remains visible.

I've changed the selected background color in 52c4f30a94ab6bc428ac981da8461470c5640301. Is it looking good?

<img width="253" alt="Screenshot 2025-05-09 at 13 41 20" src="https://github.com/user-attachments/assets/0c3adafe-cd2a-4dc0-ab20-ad4e64c1398d" />

chihsuan commented:
> I've noticed this introduces horizontal scroll with long field labels:

We could set a max-width to the label element to prevent horizontal scrolling with long field labels. Would it make sense to set a max-width of 60% for label?

jameskoster commented:
<!-- linear:replyTo:2850169950 -->

@chihsuan Maybe `--wp-components-color-accent-darker-20` would be better (sorry!). I suppose we could use the same color for the deselected checkbox border when hovering too.

<img src="https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/3381ab5e-f58d-4bbd-8a9c-f6304311c5ca/13972172-cd4a-4001-a117-ff131ab5607d?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi8zMzgxYWI1ZS1mNThkLTRiYmQtOGE5Yy1mNjMwNDMxMWM1Y2EvMTM5NzIxNzItY2Q0YS00MDAxLWExMTctZmYxMzFhYjU2MDdkIiwiaWF0IjoxNzQ2NzkzNTAzLCJleHAiOjMzMzE3MzUzNTAzfQ.qJ1m-Xy7G2q41yTO6NAwSfGSKSaDeIjOzPe020cpy-0 " alt="Screenshot 2025-05-09 at 13.24.45.png" width="858" data-linear-height="360" />

chihsuan committed:
Use --wp-components-color-accent-darker-20 for hovering color

chihsuan commented:
> Maybe `--wp-components-color-accent-darker-20` would be better (sorry!). I suppose we could use the same color for the deselected checkbox border when hovering too.

 @jameskoster Sounds good. I've updated the PR to use `--wp-components-color-accent-darker-20` for the multi-selection hover/focus state.

<img width="325" alt="Screenshot 2025-05-12 at 15 00 08" src="https://github.com/user-attachments/assets/f03a6260-2993-414f-83a1-4f46ecf764aa" />
<img width="313" alt="Screenshot 2025-05-12 at 15 00 03" src="https://github.com/user-attachments/assets/e3474f60-82b2-49d2-ac83-1516afadd0a3" />

chihsuan committed:
Merge branch 'trunk' into wooa7s-62-single-report-filters-enhance-filter-summary-component-style

youknowriad commented on packages/dataviews/src/components/dataviews-filters/style.scss:
```diff
@@ -207,6 +211,81 @@
+				}
+			}
+		}
+	}
```
There's a lot of CSS variables used here. Do we know where these CSS variables are being defined? Is it all in `@wordpress/components`? It is ok for the dataviews package to rely on wordpress/components but some of these might be considered internal APIs for now.

cc @ciampo for confirmation.

chihsuan replied:
>  Do we know where these CSS variables are being defined? Is it all in @wordpress/components?

`--wp-components-color-*` variables are all defined in `@wordpress/components`. They are provided by the [`<Theme>`](
https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/theme/styles.ts#L12-L31) component.

However, I found that it's not being used in production. It seems that it's only available in the @wordpress/components storybook. I'll update the PR to remove the references to these variables.

For the other CSS variables:

- `--wp-admin-theme-color-*` variables are defined in Core
- Other CSS variables like `--checkmark-size` are defined locally within the same file.

chihsuan replied:
Updated in 4bcbb9979b933f4f2454e39b07727c90875dbfd1.

ciampo replied:
Excuses for the delay in answering, I've been catching up on pings after some very busy weeks.

As Riad suggests, `--wp-components-color-*` are meant as internal variables of the `@wordpress/components` package. They are not advertised as public APIs, we disocurage its use [in the `Theme` README](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/theme/README.md?plain=1#L31), and in Gutenberg we had set up [a stylelint rule](https://github.com/WordPress/gutenberg/blob/d7e0c5d872ba82d851c696cc4c67ea84368a5224/.stylelintrc.js#L23) to prevent their usage.

Could we work on a follow-up PR to remove usage of these variables? (cc @mirka)

chihsuan replied:
Thanks for checking on this.

The `--wp-components-color-*` variables were already removed in commit 4bcbb99 prior to the PR being merged. Let me know if anything is missing.

ciampo replied:
Looking at the merged code, I can still see instances of those variables in `packages/dataviews/src/components/dataviews-filters/style.scss` — those are the ones I was referring to

chihsuan replied:
Thanks for catching this. 🙏  I've opened a new PR to remove them. https://github.com/Automattic/wp-calypso/pull/103615

jameskoster committed:
style adjustments

jameskoster commented:
<!-- linear:replyTo:2850169950 -->

@chihsuan I pushed a couple of small style tweaks, hope that's okay. I fixed the dimensions of the 'checkbox' and 'radio' elements, set the height of each list item to be a minimum of `32px`, and adjusted the radio hover appearance slightly to better align with the checkbox.

chihsuan commented:
> @chihsuan I pushed a couple of small style tweaks, hope that's okay. I fixed the dimensions of the 'checkbox' and 'radio' elements, set the height of each list item to be a minimum of `32px`, and adjusted the radio hover appearance slightly to better align with the checkbox.

@jameskoster That's great, thank you! 🙂

chihsuan committed:
Refactor dataviews-filters styles to use consistent theme colors for selected states

youknowriad reviewed (approved):
Let's document this change in the changelog and give it a try.

chihsuan committed:
Add changelog

chihsuan commented:
Added changelog in 7f334ce0fa6d55caf63b7b76bb195f88a9697d98

Thank you all for your feedback.
